sdk/core/Azure.Core/src/Response.cs

```diff
@@ -41,6 +41,42 @@ public abstract class Response : IDisposable
         /// </summary>
         public virtual ResponseHeaders Headers => new ResponseHeaders(this);
 
+        private BinaryData? _content;
+
+        /// <summary>
+        /// Gets the contents of HTTP response. Returns a BinaryData with no payload for responses without content.
+        /// </summary>
+        public virtual BinaryData Content
+        {
+            get
+            {
+                if (_content != null)
+                {
+                    return _content;
+                }
+
+                if (ContentStream == null)
+                {
+                    _content = new BinaryData(Array.Empty<byte>());
+                    return _content;
+                }
+
+                // If the underlying stream is a memory stream, we can just share the underlying buffer
+                // with the `BinaryData` we are going to return. Otherwise, we have to read the entire
+                // response into memory.
+                MemoryStream? memoryContent = ContentStream as MemoryStream;
+
+                if (memoryContent == null)
+                {
+                    memoryContent = new MemoryStream();
```

Comment:
Do folks think it is worthwhile to inspect the `Content-Length` header here to try to pre-size the buffer, or in cases where we have a content length will the entire response have already been read into a MemoryStream?

Reply:
>, or in cases where we have a content length will the entire response have already been read into a MemoryStream?

No, storage download responses have content length but are streamed. I think you should only create binary data if the stream is already a memory stream.

Reply:
> I think you should only create binary data if the stream is already a memory stream.

Just to confirm my understanding of the proposal: This would mean that we would make the `Content` property nullable and return `null` in these cases?

Reply:
The conversation I took part in decided to throw from Content for non-buffered responses. It seems there was another one that decided to return empty BD instead. What we definitely shouldn't do is synchronously read from live network streams.